### PR TITLE
kernel: Add device linking API

### PR DESCRIPTION
**Pull request by Sizurka**
This adds an API for creating links between devices (really between a device and a label for another device).  This arose from a [request](https://github.com/zephyrproject-rtos/zephyr/pull/15585#discussion_r286867684) by @pizi-nordic related to how to attach clock controllers/providers to devices.

Here the idea is generalized into a link type rather than being clocks only.  Currently only a clock and DMA types are defined, since those are the two I've come across in my own usage.  However, it could also be extended to provide another means to implement "parent" devices (e.g. an alternative to changing to device definition macros like #13129 is doing: add a new link type and link devices to their parent that way).

For a concrete example of how this would be used, see Sizurka/zephyr#5 (which would be merged into #15585 if this API is accepted).  Simplified, it's just adding a macro like:
```C
DEVICE_LINK_CLOCK(sam0_eic, DT_ATMEL_SAM0_EIC_0_CLOCK_CONTROLLER, 0);
```
After the device definition macro, then using code like:
```C
struct device *clk = device_get_clock(dev, 0);
```
To access the clock.

### CODE REVIEW

**Comment by pizi-nordic** [comment]
I fully support tracking device dependencies, however I think that generation of the required information should be fully automatic (all information we need is in the dts files).

As far as I know @anangl has a small script extracting all the needed information. IMHO if we combine that with your solution (providing API to get for example clocks etc.) we will end having something which satisfy all the needs.

Also, the framework should be a bit more generic in my opinion. The devices might reference not only  DMA channels, Clocks. There also could be some power regulators providing supply to part of the system as well as GPIOs or some vendor-specific things.

**Review by tbursztyka — CHANGES_REQUESTED** [review]
It's another approach than the pm's parent idea.
It also could avoid the current solution of having a dev name in the device config  (and a dev pointer in its runtime data which would be populated at init phase, but that would have to stay in order to avoid the for() loops etc..)

File: include/device.h [hunk 1]
```diff
@@ -389,6 +389,109 @@ void device_busy_set(struct device *busy_dev);
...
+ * @param dev the device linking to another one
+ * @param link_to the link target
+ * @param link_type the type of the link
+ * @param link_index the index within the same link type
+ */
+struct device_link {
+	const struct device *dev;
+	const char *link_to;
+	enum device_link_type link_type;
+	u32_t link_index;
```
[the remark this answers is not in the document]
Comment (Sizurka) [reply include/device.h]: I have no problem with this, but I'd expect the entire reason for this (instead of simply storing the linked name in the device config) would be so something that doesn't necessarily know the vendor specific ordering can access it.  

For example, without a type field a clock tree walk would have to know what each device has in each of it's fields (e.g. one might have two clocks and DMA handler, another might only have a single clock).  That kind of use would likely require an additional API to access the stored information, however.

Basically, from I'm not really sure what's to be gained if you don't differentiate the type vs storing the clock/DMA/vendor specific in the device config.

**Comment by Sizurka** [comment]
@pizi-nordic if there's already something planned for this, I have no problem with this being closed.  If there's already something planned to extract/encode phandles directly from the DTS then this is obsolete.

File: include/device.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (tbursztyka) [reply include/device.h]: Why would you do a clock tree walk at runtime? (if there is a device tree to "walk" through: it's via DTS itself, never at runtime)

About the type: one always know what device "type" it is dealing with. (you don't call spi API with a watchdog device for instance,). At runtime there is no need to know the type of a the device, since all is built-in.

[the remark this answers is not in the document]
Comment (Sizurka) [reply include/device.h]: Honestly?  I'd have to leave that to someone else to really defend, my examples are all contrived, that's beyond my current use cases.  The best case I could come up with would be controlling clocks/PM.  e.g. not all clocks are available in all sleep modes, so you either need to make the driver aware of what levels its clocks come from (extending the clock control API somehow) or allow the PM to walk the devices to figure out what ones are blocking the deeper sleeps based on what clocks they source; for SAM0 this is configurable, so the driver can't know what sleep levels it can operate in without introspecting the clock beyond what the current clock API would allow.

Regarding the device type, I'd expect it to be less about using the device specific API and more about using "global" features (e.g. PM/clock control) that don't necessarily depend on a specific device type (i.e. usage external to the driver implementation).

Either way, I'm not seeing a gain to an external API if the lookup is opaque: it seems like you'd just be better off storing it in the device config and never exposing it externally.

**Comment by pizi-nordic** [comment]
@Sizurka: So far there is no solid plan. We have considered few approaches to solve device dependency tracking problem, but no one was really complete. For example:

- The https://github.com/zephyrproject-rtos/zephyr/pull/13129 attempted to add reference to parent device (without any other information). As you can see, the proposed solution assumed heavy changes in drivers and was rejected.
- @anangl created https://github.com/anangl/zephyr/commits/extract_node_dependencies, which extracts all devices referenced by given node and places them in an array. The first entry in the array is parent, the others are random, so for example you cannot easily get all clock references.
- You created another approach, which has some benefits (for example alibity get all the clock nodes) as well as some drawbacks (requirement of drivers modification).

As you see, there is no ideal solution and every one involved looks on the problem from different perspective. IMHO we all should sit down and create something which will fulfil all the needs.

**Comment by Sizurka** [comment]
From the initial reviews, what I'm seeing is that the device linking should be both arbitrarily extensible with vendor/driver specific things without an API update and be opaque in how they are accessed (i.e. only really known the the driver implementation anyway).

To me that reads like a description of the existing strategy of storing the device name in the device config; that's already arbitrarily extensible with no outside changes and is already opaque.  

The other goal seems to be generic dependency tracking rather than specific associations, so this is overkill for that purpose.  In that case there's no need for any sort of ordering or indexing.

So I'm going to close this, since it seems like the existing way of doing it is already suitable (for specific associations) or overkill (for generic dependency tracking).
